WIP: parentheses indent and alignment

b4n opened the pull request:
This is something I started ages ago, but mentioning it in the other PR made me fix the few annoying things I had left to do :)

The last changes probably require more testing, as I just wrote them today.  The basis however should be quite OK as I am using them for quite some time now.

It's not cleaned up either or anything it's literally the steps I made to get the thing working.  Testing and comments are welcome though :)

One thing that is probably wrong is that some partial feature is enabled in "current chars" indent mode, but it seems to be more annoying than interesting when it doesn't do the whole "matching braces" logic.  So maybe I should just make it work only in "match braces" mode -- which is the only one I use anyway, so the only one actually tested.

Anyway, at least other people can now test this if they are interested :)

b4n commented:
Fixed a bug where nested alignment would not have used proper alignment characters but plain indentation (invisible of course when using spaces to indent).

```C
	foo(1,
	    bar(1,
	        2), /* here alignment assumed previous line was indented, so would
	             * have only aligned relative to bar().
	             * now it copies L2's exact indentation, including alignment */
	    3); /* here alignment was using tabs and not spaces.
	         * now, it copies L1 indent and adds alignment. */
	
```

elextr commented:
+10000 on spaces, make it the default indent setting so I don't have to change it every time I use a new config

This seems rather a big change just for alignment inside parens.  

Does it only work on functions or any parenthesised stuff eg

```C++
if ( foo( two )
     and blah () ){
    code code
}
for( int i = 0
      ; a_really_long_condition_which
        and is_continued and str[i] != ')'
      ; ++i )bletch()
```
Stupid github, the ;s are meant to align with `int` :(

Note the extra spaces and things trying to confuse it :)

@codebrainz gets many more marks for simplicity and because this has no documentation of a complex piece of software that has at least one option ;-P

But cleaned up (and when I understand what its intended to be doing) it is potentially a good idea.

b4n commented:
> This seems rather a big change just for alignment inside parens. 

As said this is definitely not cleaned up, so it might be possible to simplify it.  Though, simple alignment might look relatively easy but just look at some fixes, they ain't all that obvious :)

> Does it only work on functions or any parenthesised stuff eg
> 
> […]
>
> Note the extra spaces and things trying to confuse it :)

It aligns parentheses, wherever they are, so functions, ifs, etc.  But your extra space won't be inserted automatically with the current state -- but it should be relatively easy to implement.

> @codebrainz gets many more marks for simplicity and because this has no documentation of a complex piece of software that has at least one option ;-P

I don't get that one, what are you talking about?

> (and when I understand what its intended to be doing)

Well, the idea is "align on parentheses without getting in the way".  And apart the indentation vs. alignment it's working for me for quite some time :)


![align](https://cloud.githubusercontent.com/assets/793526/8220382/80a7774e-1552-11e5-8037-c4790b4b370b.gif)

elextr commented:
> > @codebrainz gets many more marks for simplicity and because this has no documentation of a complex piece of software that has at least one option ;-P

> I don't get that one, what are you talking about?

I was referring to #512 which is much simpler *and was documented*, was hoping it set a new standard ;)

> It aligns parentheses, wherever they are, so functions, ifs, etc. But your extra space won't be inserted automatically with the current state -- but it should be relatively easy to implement.

Yeah, just align to the first non-whitespace after the unmatched ( presuming thats the space you are talking about, the one before the `and` in the for condition is clearly not part of this PR.

I didn't read the code yet, but I assume it isn't confused by parens in strings and comments?

b4n commented:
> Yeah, just align to the first non-whitespace after the unmatched ( presuming thats the space you are talking about, the one before the `and` in the for condition is clearly not part of this PR.

Yeah.  Just did it quick&dirty in 351826258ee515d247ad06df50445e5ffb380327, seems to work :)

> I didn't read the code yet, but I assume it isn't confused by parens in strings and comments?

No, or at least it shouldn't be.

elextr commented on src/editor.c:
```diff
@@ -1352,6 +1356,150 @@ static gint get_sci_line_code_end_position(ScintillaObject *sci, gint line)
+		 * somebody wants it */
+		if (end_pos < sci_get_line_end_position (sci, match_line) - 1)
+		{
+			ret = sci_get_col_from_position(sci, end_pos + 1);
```
Obscure but ... sci_get_col is wrong if there is a double width character on the line, eg in a comment or string.

b4n replied:
Err, dammit.  Well, it's even worse than this, considering very weird cases like this

```
̂e ( dd
     bar
     baz )

ñ hello ( ff
           gg
           hh )
```

Where on the first line, display is a bit random.  Scintilla, GEdit and GVim all display the diacritic separately (as it doesn't apply to anything), using a whole character width -- so the alignment is visually correct there --, but apparently Firefox doesn't.

The ñ OTOH is always properly considered one single width visually, but indeed Scintilla reports one column too far.

For now, I'm willing to ignore this issue if Scintilla doesn't have a way to report this, as it's highly unlikely and very tricky to fix :)

elextr replied:
I was thinking of:

```C
f( "̂eñᅘ", ( some && long &&
             condition
```

but yeah, maybe its too hard for now.

Chrome displays ok, so its a firevermin problem :)

elextr commented on src/editor.c:
```diff
@@ -1352,6 +1356,150 @@ static gint get_sci_line_code_end_position(ScintillaObject *sci, gint line)
+			ret = sci_get_col_from_position(sci, end_pos + 1);
+			/* and then keep the smallest indent of a line between match and current,
+			 * not to restore a user reduced indent (e.g. for base_indent+1 style) */
+			for (++match_line; match_line < line; ++match_line)
```
May not seem intuitively needed, but may give a more useful indentation than anything else, then the user can adjust it.

b4n replied:
Yeah the idea is that if the user unindented something in the middle we shouldn't try to be more clever that (s)he was.

E.g. without it it would get this wrong:
```c
foo (
	bar (1,
	  bar (2,
	       3),
	     4 // here indent is most likely wrong, although it's aligned with the open paren
	)
)
```
With it, you get what's probably expected:
```c
foo (
	bar (1,
	  bar (2,
	       3),
	  4 // here indent is most likely correct
	)
)
```

elextr replied:
Well, I would have said the first is right. sure the user undented `bar(` likely to avoid the line getting too long, but I wouldn't have expected or wanted the next parameter to align with that, at least not automagically.

b4n replied:
Well, I find it weird to break alignment if you don't want it for real, and I really find the first example unreadable.
but if it's a common use case, maybe it can be controlled by an option…

I'm not totally sure why I implemented this in the first place (as I doubt I would have actually wrote something needing this willingly myself [1]), but I guess I encountered a situation where I found it was getting in the way of custom indent.

This said, reading the thing a bit more makes me think while the behavior is good in this particular case, the logic is flawed and simply too complex.  What it does here is find the previous unbalanced parenthesis and align to it.  But I guess a more sensible (and actually simpler and cheaper) behavior would be to stop searching whenever we find a balanced line, and use that line's indentation.
This would give this:
```C
  one(1,
    two(2, // manual unindent
        3),
    three(4, // automatically uses L2 indent
          5),
      four(6, // manual reindent
           7),
      five(8, // automatically uses L6 indent
           6)
      );
```
And this also matches better the logic of the other parts.
But with this would remove all code needed to restore manually broken alignment that you seem to want.

---
[1]:  as it's only useful if the parenthesis to match is not the last one on its line (as there is a special case not to create an alignment of the width of the previous line), so something like this doesn't need it:
```c
foo (
  1,
  bar (
    2,
    3
  )
);

b4n replied:
I implemented the simplified logic I was talking about in b476afab45645bc8e7ed70621201a46a7e4d1320 (we can always revert if we really want to realigning)

elextr commented on src/editor.c:
```diff
@@ -1352,6 +1356,157 @@ static gint get_sci_line_code_end_position(ScintillaObject *sci, gint line)
+}
+
+
+/* returns the real indentation width, not the depth */
```
Isn't this the third function with that description? :)

b4n replied:
Looks like it, it might require some cleaning up :)

elextr commented on src/editor.c:
```diff
@@ -1404,20 +1559,34 @@ static gint get_xml_indent(ScintillaObject *sci, gint line)
+/* returns whether the indentation *level* changed.
+ * new sizes and alignment are returned in @p size_ and @p align */
+static gboolean get_indent_size_after_line(GeanyEditor *editor, gint line,
+		gint *size_, gint *align, gint *line_matched)
```
`size_` YUK!!! call it `indent_size` or something

b4n replied:
trailing underscore seem a relatively common practice for output parameters, but yeah it could probably get a better name

b4n commented:
I probably should try and write some kind of test case (even simply a file with examples and expectancies) so I can at least remember why some specific logic is here… 2010 was a long time ago ^^

elextr commented:
> I probably should try and write some kind of test case (even simply a file with examples and expectancies) so I can at least remember why some specific logic is here… 2010 was a long time ago ^^

A great idea, even without the early onset Alzhiemers :)

Having a library of cases and expected results allows us to discuss it easier.

One situation to consider is case:

```C
fudge( 1, 2*
          3, // user alignment to continue expression
       4 // should align to 1 not 3
```

b4n commented:
> One situation to consider is case:
> 
> ```C
> fudge( 1, 2*
>           3, // user alignment to continue expression
>        4 // should align to 1 not 3
> ```

I don't think we (I) want to support this, because it means aligning on parentheses even when the balance didn't change -- e.g. on *each* newline.  While it can be somewhat nice (but seldom actually useful) if it does *exactly* what you want, it will become a nightmare if there is one single line it doesn't get the way you want (as it will continue on trying to wrongly align, instead of following your "fix").

And if really doing this, how not to be annoying with something like this?:

```C
fudge( 1, 2*
          3* // user alignment to continue expression
          4, // should align to 3
       5 // should align to 1 not 4
```
Of course, it'd be *possible* to do what you want, but that would mean understanding the syntax way further than parentheses alignment (it would need to somehow align on (some) commas too in this case).

IMO, if you want alignment assistance here, just add parentheses:
```C
fudge( 1, ( 2*
            3*
            4 ),
       5
```
and all is automatic (and more explicit to the reader, too).

elextr commented:
> IMO, if you want alignment assistance here, just add parentheses:
fudge( 1, ( 2*
            3*
            4 ),
       5
and all is automatic (and more explicit to the reader, too).

Good workaround, the expression is probably fairly complex if its broken over several lines so () is natural.  So you now have one case and expected solution for your list :)

b4n commented:
Rebased on master fixing merge conflicts.

Also added 2 style fixes (including one from one of @elextr's comments), and checks not to perform parentheses alignment in markup languages or the NONE filetype.

elextr commented:
Where is my documentation? And the test file noted above? :)

b4n commented:
Missing for the moment, but I'll try and work on that tomorrow, thanks for the reminder

bloer commented:
Is this still active, or any intent to merge it into mainline? I would really like this feature!
